libs/clients/zebpay/client.go

```diff
@@ -248,25 +253,33 @@ func affixAccessToken(req *http.Request, opts *ClientOpts) error {
 
 // CheckTransfer uploads the bulk payout for gemini
 func (c *HTTPClient) CheckTransfer(ctx context.Context, opts *ClientOpts, id uuid.UUID) (*CheckTransferResponse, error) {
+	resp := &CheckTransferResponse{ID: id}
 	if opts == nil {
-		return nil, errBadClientOpts
+		resp.Error = errBadClientOpts.Error()
+		return resp, errBadClientOpts
 	}
 
 	// /api/checktransferstatus/<UUID>/status
 
 	urlPath := fmt.Sprintf("/api/checktransferstatus/%s/status", id.String())
 	req, err := c.client.NewRequest(ctx, "GET", urlPath, nil, nil)
 	if err != nil {
-		return nil, err
+		resp.Error = err.Error()
+		return resp, err
 	}
 
 	// populate the access token
 	affixAccessToken(req, opts)
 
-	var resp = new(CheckTransferResponse)
 	_, err = c.client.Do(ctx, req, resp)
 	if err != nil {
-		return nil, fmt.Errorf("error %v encountered when checking status. response: %v", err, resp)
+		resp.Error = err.Error()
```

Comment:
from looking at the client code it seems that a cleaner what to do this would be to save the first return value ( which is a `*http.Response` ) and if there is an error, check if it's non-nil. if non-nil then the status code can be retrieved and used.

as per std lib http docs:
> A non-nil Response with a non-nil error only occurs when CheckRedirect fails, and even then the returned [Response.Body] is already closed.

---

as an aside I'm generally not a fan of how the error handling here works currently. would it not make more sense to handle all the code mapping here so that we may cleanly map to an enum of statuses and only return error if the http request fails? non-blocking but we should think about future refactoring

Reply:
Ah nice find. Much better.